### PR TITLE
[#8][#16][#23] Training, Deployment and Architecture sections

### LINKED ISSUE #8: Model Deployment documentation

### LINKED ISSUE #16: Unique name of API entities
The odahu tutorial uses the default names from our Terraform for odahu entities. A user can overwrite the already present connections accidentally.

### LINKED ISSUE #23: Training section

### LINKED ISSUE #35: Architecture section
We have to create a high level diagram and provide short description of out current architecture.

### PR DESCRIPTION
**Pull request by mcdoker18**
This closes #8
This closes #16 
This closes #23 
This closes #35

### CODE REVIEW

File: source/ref_deployments.rst [hunk 1]
```diff
@@ -0,0 +1,324 @@
+######################
+Model Deployments
+######################
+
+Odahu-flow Model Deployment API allows deploy ML models in a Kubernetes cluster.
```
Comment (karbyshevds) [inline source/ref_deployments.rst]:
```suggestion
Odahu-flow Model Deployment API allows deploy of ML models in a Kubernetes cluster.
```

File: source/ref_deployments.rst [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+***************************
+Model Deployment management
+***************************
+
+Model Deployments can be managed using the following ways.
+
+Swagger UI
+----------
+
+ModelDeployments are available on the Swagger UI at http://api-service/swagger/index.html URL.
```
Comment (karbyshevds) [inline source/ref_deployments.rst]:
```suggestion
ModelDeployments are available in the Swagger UI at http://api-service/swagger/index.html URL.
```

File: source/ref_deployments.rst [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+Odahu-flow CLI
+--------------
+
+:ref:`ref_odahuflowctl:Odahuflowctl` supports the Model Deployment API.
+You must be :ref:`logged in <ref_odahuflowctl:Login>` if you want to get access to the API.
```
Comment (karbyshevds) [inline source/ref_deployments.rst]:
```suggestion
You must be :ref:`logged in <ref_odahuflowctl:Login>` if you want to access the API.
```

File: source/ref_deployments.rst [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+.. code-block:: bash
+
+    odahuflowctl deployment --help
+
+* All model deployments commands and documentation:
+
+.. code-block:: bash
+
+    odahuflowctl deployment --help
+
```
Comment (karbyshevds) [inline source/ref_deployments.rst:90]: is it duplicated section?

File: source/ref_deployments.rst [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+********************************************
+Feedback
+********************************************
+
+**Model Feedback** provides a view of performance over all stages of model lifecycle.
+
+The mechanism is simple:
+
+1. Ask Deploy for prediction (with or without ``Request-Id`` provided)
+2. Send Feedback to Odahu-flow about the prediction (with ``Request-Id`` returned from previous step)
```
Comment (karbyshevds) [inline source/ref_deployments.rst]:
```suggestion
2. Send prediction feedback to Odahu-flow  (with ``Request-Id`` returned from previous step)
```

File: source/ref_deployments.rst [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+3. Odahu-flow stores the prediction and feedback to a configurable location
+
+.. important::
+
+   This flow requires ``feedback`` to be enabled in ``values.yaml`` during Helm chart install
```
Comment (karbyshevds) [inline source/ref_deployments.rst]:
```suggestion
   This flow requires ``feedback`` to be enabled in ``values.yaml`` during Helm chart installation
```

File: source/ref_deployments.rst [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+Protocol
+--------
+
+1. If prediction is requested without Request-ID: ``Request-ID`` header with random ID is added to the request. Otherwise, Request-ID is not generated.
+2. Request and response are stored on configured external storage (eg. S3, GCS)
+3. User sends Model Feedback as an argument to the feedback endpoint. (Feedback can be arbitrary JSON.)
+5. All Feedback is persisted on external storage and can be used by models during subsequent Trains.
+
+Worked example
```
Comment (karbyshevds) [inline source/ref_deployments.rst]: Should it be `working example` or `work example` ?

File: source/ref_trainings.rst [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+Data source supports the following types of Odahu-flow connections:
+
+    * :ref:`ref_connections:S3`
+    * :ref:`ref_connections:Google Cloud Storage`
+    * :ref:`ref_connections:Azure Blob storage`
+
+Let's consider the following example of downloading training data from Google Cloud Storage.
+
+Prerequisites:
+    * The training data set locates in the `wine-training-data` bucket by `wine/11-11-2011/` directory.
```
Comment (karbyshevds) [inline source/ref_trainings.rst]:
```suggestion
    * The training data set is located in the `wine-training-data` bucket by `wine/11-11-2011/` directory.
```

File: source/ref_trainings.rst [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    * The ML script expects that the data will locate in `data/` directory relative to the root git directory.
```
Comment (karbyshevds) [inline source/ref_trainings.rst]:
```suggestion
    * The ML script expects that the data will be located in `data/` directory relative to the root git directory.
```

File: source/ref_trainings.rst [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+    odahuflowctl train get --format json
+
+Getting the model name of the trainings:
+
+.. code-block:: bash
+
+    odahuflowctl train get --id tensorflow-cli -o 'jsonpath=[*].spec.model.name'
+
+* Creating of a training from `train.yaml` file:
```
Comment (karbyshevds) [inline source/ref_trainings.rst]:
```suggestion
* Creating a training from `train.yaml` file:
```

File: source/ref_trainings.rst [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        parameters:
+          data_file: path
+          regularization: {type: float, default: 0.1}
+        command: "python train.py -r {regularization} {data_file}"
+      test:
+        parameters:
+          data_file: path
+        command: "python validate.py {data_file}"
+
+Model Training API only can contain one entry point.
```
Comment (karbyshevds) [inline source/ref_trainings.rst]:
```suggestion
Model Training API can contain only one entry point.
```

File: source/ref_trainings.rst [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+
+Odahu-flow requires that a model be logged through `mlflow API <https://www.mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.log_model>`_.
+
+.. code-block:: python
+    :caption: Example of sklearn model logging:
+
+    mlflow.sklearn.log_model(lr, "model")
+
+Optionally, you can provide input and output samples for Odahu-flow.
+It allows determining input and output types for Odahu-flow packagers.
```
Comment (karbyshevds) [inline source/ref_trainings.rst:260]:
```suggestion
It allows to determine input and output types for Odahu-flow packagers.
```

**Review by karbyshevds — CHANGES_REQUESTED** [review]

File: source/ref_trainings.rst [hunk 13]
```diff
@@ -0,0 +1,269 @@
+######################
+Model Trainings
```
Comment (vlad-tokarev) [inline source/ref_trainings.rst:2]: Please rename it as just `Trainings` and `Model Deployments` as  just `Deployments`
OR
Rename `Packagers` to `Model Packagers` 

for better consistency.

**Review by vlad-tokarev — APPROVED** [review]

File: source/ref_deployments.rst (hunk 4 above)
Comment (mcdoker18) [reply source/ref_deployments.rst:90]: Noop.  All APIs have the same section.

File: source/ref_trainings.rst (hunk 12 above)
Comment (mcdoker18) [reply source/ref_trainings.rst:260]: Grammarly recommends the gerund form in the this case.

File: source/ref_deployments.rst (hunk 7 above)
Comment (mcdoker18) [reply source/ref_deployments.rst]: Fixed

File: source/ref_deployments.rst [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+Example of a prediction request:
+
+.. image:: img/service_catalog_predict.png
+
+********************************************
+Grafana Dashboard
+********************************************
+
+Out of the box, Odahu-flow provides the Grafana Model Deploymet dashboard.
```
Comment (maxkochubey) [inline source/ref_deployments.rst]:
```suggestion
Out of the box, Odahu-flow provides the Grafana Model Deployment dashboard.
```

File: source/ref_deployments.rst [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+      "model_version": "1.0",
+      "request_http_headers": {
+        ":method": "POST",
+        "x-scheme": "https",
+        "accept": "*/*",
+        "x-istio-attributes": "ClEKF2Rlc3RpbmF0aW9uLnNlcnZpY2UudWlkEjYSNGlzdGlvOi8vb2RhaHUtZmxvdy1kZXBsb3ltZW50L3NlcnZpY2VzL3Rlc3QtZTJlLXdpbmUKUwoYZGVzdGluYXRpb24uc2VydmljZS5ob3N0EjcSNXRlc3QtZTJlLXdpbmUub2RhaHUtZmxvdy1kZXBsb3ltZW50LnN2Yy5jbHVzdGVyLmxvY2FsCisKGGRlc3RpbmF0aW9uLnNlcnZpY2UubmFtZRIPEg10ZXN0LWUyZS13aW5lCjgKHWRlc3RpbmF0aW9uLnNlcnZpY2UubmFtZXNwYWNlEhcSFW9kYWh1LWZsb3ctZGVwbG95bWVudApPCgpzb3VyY2UudWlkEkESP2t1YmVybmV0ZXM6Ly9pc3Rpby1pbmdyZXNzZ2F0ZXdheS04NjlkYjdkOWJiLWpsemtyLmlzdGlvLXN5c3RlbQ==",
+        "knative-serving-namespace": "odahu-flow-deployment",
+        "x-envoy-original-path": "/model/test-e2e-wine/api/model/invoke",
+        ":scheme": "http",
+        "x-forwarded-host": "odahu.gke-legion-dev01.ailifecycle.org",
```
Comment (maxkochubey) [inline source/ref_deployments.rst]:
```suggestion
        "x-forwarded-host": "odahu.example.com",
```

File: source/ref_deployments.rst [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+        "x-forwarded-port": "443",
+        "content-type": "application/json",
+        "x-b3-sampled": "1",
+        "content-length": "257",
+        "x-real-ip": "10.4.0.11",
+        "x-forwarded-for": "10.4.0.11,10.44.0.10",
+        "x-envoy-external-address": "10.44.0.10",
+        "x-request-id": "12dcddd5-771d-9fc0-b326-816211cf8172",
+        "x-forwarded-proto": "http",
+        ":authority": "odahu.gke-legion-dev01.ailifecycle.org",
```
Comment (maxkochubey) [inline source/ref_deployments.rst]:
```suggestion
        ":authority": "odahu.example.com",
```

File: source/ref_deployments.rst [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+        "x-b3-traceid": "488abe197a652c2ce80e6b848a4c56e6",
+        ":path": "/api/model/invoke",
+        "accept-encoding": "gzip, deflate",
+        "x-envoy-decorator-operation": "test-e2e-wine.odahu-flow-deployment.svc.cluster.local:80/model/test-e2e-wine/api*",
+        "x-b3-spanid": "e80e6b848a4c56e6",
+        "knative-serving-revision": "test-e2e-wine-l4xt7"
+      },
+      "response_status": "200",
+      "request_host": "odahu.gke-legion-dev01.ailifecycle.org",
```
Comment (maxkochubey) [inline source/ref_deployments.rst]:
```suggestion
      "request_host": "odahu.example.com",
```

**Review by maxkochubey — APPROVED** [review]

**Pull request merged**